### ISSUE TITLE
Atomic error on saving after expired lock

### ISSUE THREAD

**Issue by BrendaH** [rparent/django-lock-tokens#5]
* django-lock-tokens version: 0.1.4
* Django version: 1.11.15
* Python version: 3.5
* Operating System: Windows 10

### Description & What I Did
Since my custom change_form template does not play well with the template provided by django-lock-tokens (see #4) I am trying to set and remove locks in the view that shows and handles the form.

On top of the view I have:
```
    if check_for_session(schedule, request.session):  # Returns True if lock was not found or was expired
        try:
            lock_for_session(schedule, request.session)
        except AlreadyLockedError:
            messages.error(request, ugettext_lazy('Someone else has locked the object you want to work on, you can not work on it at the moment.'), extra_tags='bg-danger')
            return HttpResponseRedirect(reverse('admin:pe_schedule_changelist'))
```

Then some irrelevant code.
And just before redirecting after saving stuff I have:
```
if request.method == 'POST':
    if form.is_valid() and formset.is_valid():
            [ do all kinds of saving stuff ]
            try:
                unlock_for_session(workschedule, request.session)
            except UnlockForbiddenError:
                pass  # TODO: should we do anything special? No lock, no big deal, right?

            [ redirect to a different view ]
// when not valid, no unlocking just reshowing the form
// when not post, no unlocking just showing the form
```
So, this works rather well when I open a schedule to edit, and then save it. The lock is removed and someone else can edit the schedule.
However, when I leave the lock to expire (set it to 1 minute) while still in the editing interface and then saving, I get a bunch of errors.
Looks like something tries to get() a specific lock but this fails and is not caught. I think this is where a query is killed and the atomic thing isn't correctly closed.

Anybody got any idea how I should go about creating and deleting locks? How can I allow saving without errors when a lock is already expired?

```
Traceback:

File "C:\[path to virtualenv]\lib\site-packages\lock_tokens\managers.py" in get_or_create_for_object
  17.             return (self.get_for_object(obj), False)

File "C:\[path to virtualenv]\lib\site-packages\lock_tokens\managers.py" in get_for_object
  13.                         locked_at__gte=get_oldest_valid_tokens_datetime())

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\manager.py" in manager_method
  85.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\query.py" in get
  380.                 self.model._meta.object_name

During handling of the above exception (LockToken matching query does not exist.), another exception occurred:

File "C:\[path to virtualenv]\lib\site-packages\django\db\backends\utils.py" in execute
  64.                 return self.cursor.execute(sql, params)

File "C:\[path to virtualenv]\lib\site-packages\django\db\backends\sqlite3\base.py" in execute
  328.         return Database.Cursor.execute(self, query, params)

The above exception (UNIQUE constraint failed: lock_tokens_locktoken.locked_object_content_type_id, lock_tokens_locktoken.locked_object_id) was the direct cause of the following exception:

File "C:\[path to virtualenv]\lib\site-packages\lock_tokens\models.py" in save
  59.             return super(LockToken, self).save(*args, **opts)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\base.py" in save
  808.                        force_update=force_update, update_fields=update_fields)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\base.py" in save_base
  838.             updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\base.py" in _save_table
  924.             result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\base.py" in _do_insert
  963.                                using=using, raw=raw)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\manager.py" in manager_method
  85.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\query.py" in _insert
  1076.         return query.get_compiler(using=using).execute_sql(return_id)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\sql\compiler.py" in execute_sql
  1112.                 cursor.execute(sql, params)

File "C:\[path to virtualenv]\lib\site-packages\django\db\backends\utils.py" in execute
  79.             return super(CursorDebugWrapper, self).execute(sql, params)

File "C:\[path to virtualenv]\lib\site-packages\django\db\backends\utils.py" in execute
  64.                 return self.cursor.execute(sql, params)

File "C:\[path to virtualenv]\lib\site-packages\django\db\utils.py" in __exit__
  94.                 six.reraise(dj_exc_type, dj_exc_value, traceback)

File "C:\[path to virtualenv]\lib\site-packages\django\utils\six.py" in reraise
  685.             raise value.with_traceback(tb)

File "C:\[path to virtualenv]\lib\site-packages\django\db\backends\utils.py" in execute
  64.                 return self.cursor.execute(sql, params)

File "C:\[path to virtualenv]\lib\site-packages\django\db\backends\sqlite3\base.py" in execute
  328.         return Database.Cursor.execute(self, query, params)

During handling of the above exception (UNIQUE constraint failed: lock_tokens_locktoken.locked_object_content_type_id, lock_tokens_locktoken.locked_object_id), another exception occurred:

File "C:\[path to virtualenv]\lib\site-packages\django\core\handlers\exception.py" in inner
  41.             response = get_response(request)

File "C:\[path to virtualenv]\lib\site-packages\django\core\handlers\base.py" in _get_response
  187.                 response = self.process_exception_by_middleware(e, request)

File "C:\[path to virtualenv]\lib\site-packages\django\core\handlers\base.py" in _get_response
  185.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "C:\[path to virtualenv]\lib\site-packages\django\contrib\admin\options.py" in wrapper
  552.                 return self.admin_site.admin_view(view)(*args, **kwargs)

File "C:\[path to virtualenv]\lib\site-packages\django\utils\decorators.py" in _wrapped_view
  149.                     response = view_func(request, *args, **kwargs)

File "C:\[path to virtualenv]\lib\site-packages\django\views\decorators\cache.py" in _wrapped_view_func
  57.         response = view_func(request, *args, **kwargs)

File "C:\[path to virtualenv]\lib\site-packages\django\contrib\admin\sites.py" in inner
  224.             return view(request, *args, **kwargs)

File "C:\[path to virtualenv]\lib\site-packages\reversion\admin.py" in change_view
  180.             return super(VersionAdmin, self).change_view(request, object_id, form_url, extra_context)

File "C:\[path to virtualenv]\lib\site-packages\django\contrib\admin\options.py" in change_view
  1512.         return self.changeform_view(request, object_id, form_url, extra_context)

File "C:\_D_\Werk\Real Life Gaming\Community portal\complatform\pe\admin.py" in changeform_view
  439.         return fill_schedule(request, self, object_id, extra_context)

File "C:\[path to virtualenv]\lib\site-packages\django\contrib\auth\decorators.py" in _wrapped_view
  23.                 return view_func(request, *args, **kwargs)

File "C:\_D_\Werk\Real Life Gaming\Community portal\complatform\pe\adminviews.py" in fill_schedule
  1284.             lock_for_session(schedule, request.session)

File "C:\[path to virtualenv]\lib\site-packages\lock_tokens\sessions.py" in lock_for_session
  14.     lock_token = LockableModel.lock(obj, token)

File "C:\[path to virtualenv]\lib\site-packages\lock_tokens\models.py" in lock
  99.         lock_token = self._lock(token)

File "C:\[path to virtualenv]\lib\site-packages\lock_tokens\models.py" in _lock
  78.         lock_token, created = LockToken.objects.get_or_create_for_object(self)

File "C:\[path to virtualenv]\lib\site-packages\lock_tokens\managers.py" in get_or_create_for_object
  19.             return (self.create(locked_object=obj), True)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\manager.py" in manager_method
  85.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\query.py" in create
  394.         obj.save(force_insert=True, using=self.db)

File "C:\[path to virtualenv]\lib\site-packages\lock_tokens\models.py" in save
  63.                                           locked_object_content_type=self.locked_object_content_type)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\manager.py" in manager_method
  85.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\query.py" in get
  374.         num = len(clone)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\query.py" in __len__
  232.         self._fetch_all()

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\query.py" in _fetch_all
  1118.             self._result_cache = list(self._iterable_class(self))

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\query.py" in __iter__
  53.         results = compiler.execute_sql(chunked_fetch=self.chunked_fetch)

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\sql\compiler.py" in execute_sql
  899.             raise original_exception

File "C:\[path to virtualenv]\lib\site-packages\django\db\models\sql\compiler.py" in execute_sql
  889.             cursor.execute(sql, params)

File "C:\[path to virtualenv]\lib\site-packages\django\db\backends\utils.py" in execute
  79.             return super(CursorDebugWrapper, self).execute(sql, params)

File "C:\[path to virtualenv]\lib\site-packages\django\db\backends\utils.py" in execute
  59.         self.db.validate_no_broken_transaction()

File "C:\[path to virtualenv]\lib\site-packages\django\db\backends\base\base.py" in validate_no_broken_transaction
  448.                 "An error occurred in the current transaction. You can't "

Exception Type: TransactionManagementError at /nl/admin/pe/schedule/42/change/
Exception Value: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
```

**Comment by BrendaH** [comment]
I think that the problem might be that the lock expires but is not deleted anywhere. When the LockableModel._check_and_get_lock_token() tries to find a valid token it is not found. Then it tries to make a new one for the same item but this is not allowed due to the unique constraint and then you get the traceback I posted before.

So either some javascript should be running to clear away old locks (but where to put that, and when to run it? Looks like it only has clear_all and that's a bit brutal), or perhaps the code that checks for a valid lock should also check for an expired lock and either renew or delete it if it is found. However, this brings complications when someone else has started editing and the person with the expired lock (and page) tries to save.

Any ideas/suggestions?

**Comment by rparent** [comment]
Are you trying to modify the views from your `ModelAdmin` class here ? Because if it inherits from `LockableModelAdmin`, the locking mechanism is already handled in the view (server side), so you shouldn't need to implement it yourself. And if you call `super()` somewhere in the code, then it could explain the exception you posted.

The only thing that is done client side in javascript (and that you don't have since you override the template) is to try to unlock the model before the user leaves the page by another way than by saving (e.g. by hitting "previous" in the navigator or by closing the tab). 

So in order to get the full mechanism to work when you override the template, you only need to re-write this javascript part somewhere (or wait for the resolution of the previous [issue](https://github.com/rparent/django-lock-tokens/issues/4) you posted :) ).

**Comment by BrendaH** [comment]
I have my own views defined for displaying the form. It's a two step process: on adding an item the admin gets a form with 3 fields, after saving that he gets the next form with the same 3 fields (filled in, read only) and quite a lot of formsets.
So in my `class ScheduleAdmin(VersionAdmin, LockableModelAdmin)` I have;
```
    # Adjust add and change buttons to go to the right, custom, view
    def add_view(self, request, form_url='', extra_context=None):
        return create_schedule(request, self, 0)

    change_form_template = 'admin/pe/schedule/change_form.html'

    # Use changeform_view instead of (similar) change_view because Reversion uses changeform_view directly
    def changeform_view(self, request, object_id=None, form_url='', extra_context=None):
        return fill_schedule(request, self, object_id, extra_context)
```
My change_form no longer extends the `admin/lock_tokens_change_form.html` so I manually include the API and added some custom javascript to handle the locks.

I am passing (via extra_context) the lock_token and registering it in the jscript API and then calling `LockTokens.hold_lock` so jscript can keep it up to date for as long as the page is open. There's now also a timeout to have the user choose if he wants to renew the lock or abandon the page and the edits after a certain amount of time. (next step is working on the case that someone doesn't see the prompt and the page is still open)

I added the `window.onunload()` as you suggested so that case is also covered.

**Comment by BrendaH** [comment]
I do inherit from `LockableModelAdmin`, however I have my own `changeform_view` method and my `fill_schedule` view handles the post and saving of stuff etc. So I skip calling `save_model`, in which `LockableModelAdmin` does some other check on the lock. However, manually calling that after my own save code does not work either.

I have been looking at this for two days now and I just can't get it to work properly....

Also, I do not understand the following:
`check_for_session` goes via `LockableModel.check_lock_token(self, token)` to `LockableModel._check_and_get_lock_token(self, token)` and one of the first things it does is trying `LockToken.objects.get_for_object(self)`.
However, I don't think that `get_for_object `is every going to find something in case of an expired lock because it is explicitly looking for a lock of within the last timeout period (1 minute in my case for faster testing). So how does `_check_and_get_lock_token` ever get to the line where it checks if the token is expired in case of an expired token?
Why not use `LockToken.objects.get(token_str=token)` inside `LockableModel._check_and_get_lock_token`, since you do have a token string, right?

However.. replacing that still does not solve my problem. :-(
Via some undefinable way `LockableModel.lock` is called (I don't know why or from where) and everything blows up in my face with that atomic warning.  :-(

**Comment by matthijskooijman** [comment]
@BrendaH asked me to have a look at this issue. I'm responding only to the original issue now. I believe the problem is that @BrendaH is running inside a database transaction, which is not compatible with the way `LockToken.save()` works. In particular this bit: https://github.com/rparent/django-lock-tokens/blob/ddb5dbc7ae5bf19d361604ffdf3c52c6d8129a1a/lock_tokens/models.py#L58-L66

Looking at the django docs on transactions ([this section](https://docs.djangoproject.com/en/2.1/topics/db/transactions/#django.db.transaction.atomic), scroll down to "Avoid catching exceptions inside atomic!"), it seems that inside an atomic transaction, you should not catch an `IntegrityError` or at least not try to do additional database transactions when handling the error. Once a database constraint has failed, the transaction has failed as well, and no database queries can happen.

To fix this (apart from requiring to outside of a transaction, which seems unreasonable to me), I can imagine simply changing the current "delete an expired token on `IntegrityError`"-approach, to a simpler "delete expired tokens before saving"-approach. If an IntegrityError still happens, there is an active lock so save can just raise an `AlreadyLockedError` (without needing further database queries). 

Intuition tells me there's a risk of race conditions, but thinking about it, I can see:
 - The case where another lock expires between the check for expired tokens and the actual save. This will give an `IntegrityError` and thus an `AlreadyLockedError`, even though the other lock *just* expired. In practice, this is ok, if the check happened just a tiny bit earlier, you would have had the same result.
 - The case where a lock is renewed after the check for expired tokens and before the deletion. This is a real race condition, but I believe it already exists now and should be solved by making the deletion, as well as the renewal atomic.

**Comment by rparent** [comment]
Hello @matthijskooijman,

Thanks for the answer. You are right about the cause of the error, and I agree that the lock mechanism shouldn't require to commit an open transaction (which is what I had in mind when first implementing this, as far as I can remember). I usually do the object locking in its own transaction without any other database operation, but it shouldn't be mandatory.

Your approach of simply deleting expired tokens before saving seems saner regarding transactions handling. I will have a look and try to fix this.

**Comment by rparent** [comment]
@BrendaH I just commited a patch that should resolve your transaction issue, based on @matthijskooijman suggestion. 
If you have a chance to test it on your project to tell me if it works, that would be great.

**Comment by matthijskooijman** [comment]
Looks good to me. I happen to know @BrendaH is busy tomorrow, but she'll probably be happy to test this on thursday :-)

**Comment by BrendaH** [comment]
@rparent Thanks! I have a full day of meetings today and after that should first finish some other business I am in the middle of at the moment. But I hope to be able to test your patch this week. :-)

**Comment by BrendaH** [comment]
@rparent I'm very sorry but I won't be able to test this week due to a fast approaching deadline and more work than expected. I really appreciate your quick responses and truly hope to find some time soon to check out your patch.

**Comment by rparent** [comment]
I just released version 0.2.1 that include the bug fix for this issue, so I'm going to mark it as closed. Feel free to reopen if you think that the bug is not fixed.
